From 26e78958b54c3e738df939af171c6f3cafc9fbf9 Mon Sep 17 00:00:00 2001
From: Mohan Raj <quic_mrajraje@quicinc.com>
Date: Tue, 11 Jun 2024 23:14:59 +0530
Subject: [PATCH] hostapd: display the link_type of a station through hostapd_cli 
 
Getting the link type status of a Station from the STA Control field
Each STA Control field contains NSTR Present subfield
which indicates if at least one NSTR is present in the
MLD that contains the link corresponding to that STA.
NSTR Present subfield is set to 1 if there is at least one NSTR
is present and otherwise, it is set to 0 which indicates the
STR mode

Usage: hostapd_cli -i <interface> -l <link> all_sta

Output:
06:1c:f9:22:14:fc
flags=[AUTH][ASSOC][AUTHORIZED][WMM][MFP][HT][VHT][HE][EHT]
aid=1
capability=0x0
listen_interval=1
supported_rates=8c 12 98 24 b0 48 60 6c fb
timeout_next=NULLFUNC POLL
vendor_oui=00:00:00
dot11RSNAStatsSTAAddress=06:1c:f9:22:14:fc
dot11RSNAStatsVersion=1
dot11RSNAStatsSelectedPairwiseCipher=00-0f-ac-4
dot11RSNAStatsTKIPLocalMICFailures=0
dot11RSNAStatsTKIPRemoteMICFailures=0
wpa=2
AKMSuiteSelector=00-0f-ac-8
hostapdWPAPTKState=11
hostapdWPAPTKGroupState=0
hostapdMFPR=1
rx_packets=261
tx_packets=8
rx_bytes=134281
tx_bytes=1814
inactive_msec=25610
signal=67
rx_rate_info=60
tx_rate_info=60
rx_vht_mcs_map=ffaa
tx_vht_mcs_map=ffaa
ht_mcs_bitmask=ffffffff000000000000
connected_time=25
supp_op_classes=80737475767778797a7b7c7d7e7f808182
uniibands supported:
        UNII_5 UNII_6 UNII_7 UNII_8
        UNII_1 UNII_2A UNII_2C UNII_3 UNII_4
        UNII_1
        UNII_2A
        UNII_2C
        UNII_3
        UNII_4
min_txpower=0
max_txpower=75
vht_caps_info=0x3383f9b2
ht_caps_info=0x09ef
ext_capab=04000f00000000400040000000
max_nss=4
max_vhtmcs=9
link_type[1]=STR

Signed-off-by: Mohan Raj <quic_mrajraje@quicinc.com>
---
 src/ap/ctrl_iface_ap.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/src/ap/ctrl_iface_ap.c b/src/ap/ctrl_iface_ap.c
index 72821b4..552b41c 100644
--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -726,6 +726,17 @@ static int hostapd_ctrl_iface_sta_mib(struct hostapd_data *hapd,
 	 }
 #endif /* CONFIG_IEEE80211N */
 
+#ifdef CONFIG_IEEE80211BE
+	if (sta->mld_info.mld_sta == true) {
+		size_t link_type;
+
+		link_type = sta->mld_info.links[hapd->mld_link_id].nstr_bitmap_len;
+		ret = os_snprintf(buf + len, buflen-len, "link_type[%d]=%s\n",
+				 hapd->mld_link_id, link_type ? "NSTR" : "STR");
+		if (!os_snprintf_error(buflen - len, ret))
+			len += ret;
+	}
+#endif /* CONFIG_IEEE80211BE*/
 	return len;
 }
 
-- 
2.34.1

